contracts/project/Builders.sol

```diff
@@ -12,26 +12,147 @@ contract Builders {
         bytes32 publicKey;
     }
 
+    struct BuilderJoinRequest {
+        address _address;
+        bool parity;
+        bytes32 publicKey;
+    }
+
     mapping(address => Builder) public builders;
+    address[] public buildersArray;
+
+    mapping(address => BuilderJoinRequest) public builderRequests;
+    address[] public builderRequestsArray;
 
-    modifier onlyBuilder {
+    modifier onlyBuilder() {
         require(
             builders[msg.sender]._address != address(0),
             "Only builders are allowed to execute this."
         );
         _;
     }
 
+    function getBuildersLength() public view returns (uint256) {
+        return buildersArray.length;
+    }
+
+    function getBuilderRequestsLength() public view returns (uint256) {
+        return builderRequestsArray.length;
+    }
+
     /**
      * @notice Builder can update his public key.
      * @param parity based on header value (0x02/0x03 - false/true)
      * @param publicKey compressed public key value
      */
     function setBuilderPublickey(bool parity, bytes32 publicKey) public {
-        require(builders[msg.sender]._address == msg.sender, "Builder not set");
+        require(
+            builders[msg.sender]._address == msg.sender,
+            "Sender is not builder"
+        );
         builders[msg.sender].parity = parity;
         builders[msg.sender].publicKey = publicKey;
     }
 
+    /**
+     * @notice Already accepted builder can add new builder.
+     * @param newBuilderAddress address of the new builder
+     * @param parity based on header value (0x02/0x03 - false/true)
+     * @param publicKey compressed public key value
+     */
+    function addBuilder(
+        address newBuilderAddress,
+        bool parity,
+        bytes32 publicKey
+    ) public onlyBuilder {
+        require(
+            builders[newBuilderAddress]._address == address(0),
+            "Builder already exists"
+        );
+
+        builders[newBuilderAddress] = Builder({
+            _address: newBuilderAddress,
```

Comment:
I think address should be `msg.sender`. We should not get it as an argument

Reply:
This is for option when builder wants to add another builder from the UI, without having to go through process of creating request (might be the possibility in the future), so `msg.sender` is already builder. When external account wants to become builder, he has to create request in `requestJoinBuilder`.

Reply:
I think we can leave it here. For example functionality of adding list of builders rather than repeat request-accept will be useful.